text/0000-ques-in-main.md

```diff
@@ -0,0 +1,514 @@
+general.  The stock `Termination` impls in libstd mean that simple
+programs that can fail don't need to do anything special:
+
+``` rust
+fn main() -> Result<(), io::Error> {
+    let mut stdin = io::stdin();
+    let mut raw_stdout = io::stdout();
+    let mut stdout = raw_stdout.lock();
+    for line in stdin.lock().lines() {
+        stdout.write(line?.trim().as_bytes())?;
+        stdout.write(b"\n")?;
+    }
+    stdout.flush()
+}
+```
+
+Doctest examples can freely use `?` with no extra boilerplate;
+`#[test]` examples may need their boilerplate adjusted.
+
+More complex programs, with their own error types, still get
+platform-consistent exit status behavior for free, as long as they
+implement `Error`.  The Book should describe `Termination` to explain
+_how_ `Result`s returned from `main` turn into error messages and exit
+statuses, but as a thing that most programs will not need to deal with
+directly.
+
+Tutorial examples should probably still begin with `fn main() -> ()`
+until the tutorial gets to the point where it starts explaining why
+`panic!` and `unwrap` are not for "normal errors".
+
+Discussion of `TermStatus` should be reserved for an advanced-topics
+section talking about interoperation with the Unix command-line
+ecosystem.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Generalizing the return type of `main` complicates libstd and/or the
+compiler.  It also adds an additional thing to remember when complete
+newbies to the language get to error handling.  On the other hand,
+people coming to Rust from other languages may find this _less_
+surprising than the status quo.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Do nothing; continue to live with the trip hazard, the extra
+boilerplate required to comply with platform conventions, and people
+using `panic!` to report ordinary errors because it's less hassle.
+
+The [pre-RFC][pre-rfc] included a suggestion to use `catch` instead,
+but this still involves extra boilerplate in `main` so I'm not
+enthusiastic about it.
+
+# Unresolved Questions
+[unresolved]: #unresolved-questions
+
+We need to decide what to call the new trait.  The names proposed in
+the pre-RFC thread were `Terminate`, which I like OK but have changed
+to `Termination` because return value traits should be nouns, and
+`Fallible`, which feels much too general, but could be OK if there
+were other uses for it?  Relatedly, it is conceivable that there are
+other uses for `Termination` in the existing standard library, but I
+can't think of any right now.  (Thread join was mentioned in the
+[pre-RFC][pre-rfc], but that can already relay anything that's `Send`,
+so I don't see that it adds value there.)
+
+We also need to decide where the trait should live.  One obvious place
+is in `std::process`, because that is where `exit(i32)` is; on the
+other hand, this is basic enough that it may make sense to put at
+least some of it in libcore.
```

Comment:
`Termination` should be put in `libcore`, probably as a lang-item. I don't think it is a good idea for the compiler to use a hard-coded list of valid return types, where the path `std::error::Error`, and worse, the OS-dependent `std::process::TermStatus` need to be known by the compiler. 

(Also note that the current `?` operator lowering does not refer to `core::result::Result` directly, but through the trait `core::ops::Carrier`.)